### PR TITLE
Klokane/refract - DO NOT MERGE

### PR DESCRIPTION
**Pull request by klokane**
@zdne @pksunkara Please go throught test/fixtures/mson-*.(apib|json) pair and check if they are serialized correctly against refract spec

### CODE REVIEW

**Comment by pksunkara** [comment]
I didn't realise you added new commits. I was in the middle of the review. Going to review this fresh again first thing tomorrow morning.

File: src/SerializeAST.cc [hunk 1]
```diff
@@ -444,8 +454,10 @@ sos::Object WrapDataStructure(const DataStructure& dataStructure)
     // Type Sections
     dataStructureObject.set(SerializeKey::Sections,
                             WrapCollection<mson::TypeSection>()(dataStructure.sections, WrapTypeSection));
+
```
Comment (pksunkara) [inline src/SerializeAST.cc]: Why this extra line?

Comment (klokane) [inline src/SerializeAST.cc]: Accident while edit

File: src/refract/Element.h [hunk 2]
```diff
@@ -0,0 +1,444 @@
+//
+//  refract/Element.h
+//  librefract
+//
+//  Created by Jiri Kratochvil on 18/05/15.
+//  Copyright (c) 2015 Apiary Inc. All rights reserved.
+//
+#if !defined _REFRACT_ELEMENT_H_
```
Comment (pksunkara) [inline src/refract/Element.h]: `#ifndef` should be used. And the symbol name should be changed too.

File: src/refract/Element.h [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+    // if you want add new one, do not forget add IElement::Visitors
+    struct ComparableVisitor;
+    struct SerializeVisitor;
+    struct SerializeCompactVisitor;
+
+    struct LogicError : std::logic_error {
+        explicit LogicError(const std::string& msg) : std::logic_error(msg) {}
+    };
+    // will be removed in future
+    struct NotImplemented : std::runtime_error {
```
Comment (pksunkara) [inline src/refract/Element.h]: Please add empty lines before new structs. Improves readability and maintainability of the code.

File: src/refract/Typelist.h [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+//  refract/Typelist.h
+//  librefract
+//
+//  Based on concept of typelist introduced in
+//  http://www.drdobbs.com/generic-programmingtypelists-and-applica/184403813
+//  
+//  Created by Jiri Kratochvil on 18/05/15.
+//  Copyright (c) 2015 Apiary Inc. All rights reserved.
+//
+#if !defined _REFRACT_TYPELIST_H_
```
Comment (pksunkara) [inline src/refract/Typelist.h]: `#ifndef`

File: src/refract/VisitableBy.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
+//
+//  refract/VisitableBy.h
+//  librefract
+//
+//  Created by Jiri Kratochvil on 18/05/15.
+//  Copyright (c) 2015 Apiary Inc. All rights reserved.
+//
+#if !defined _REFRACT_VISITABLE_BY_H_
```
Comment (pksunkara) [inline src/refract/VisitableBy.h]: `#ifndef` is the style we are using.

File: src/refract/Element.h [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+        void visit(const MemberElement& e);
+        void visit(const ObjectElement& e);
+
+        sos::Object get()
+        {
+            return result;
+        }
+    };
+
+    struct SerializeCompactVisitor : IVisitor
```
Comment (pksunkara) [inline src/refract/Element.h]: I am just wondering if all these serialize stuff should be in `Element.h` itself. Can we use another header file?

File: test/test-RefractTest.cc [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+#include <string>
+
+#include "snowcrash.h"
+
+#include "sosJSON.h"
+#include "SerializeAST.h"
+
+
+bool HandleFixtureTest(const std::string& basepath) {
```
Comment (pksunkara) [inline test/test-RefractTest.cc]: `{` should be on a new line in our `.cc` files.

File: test/fixtures/mson-enum.json [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+      "content": [
+        {
+          "element": "object",
+          "meta": {
+            "id": "User",
+            "title": "User"
+          },
+          "content": [
+            {
+              "element": "object",
```
Comment (pksunkara) [inline test/fixtures/mson-enum.json]: I don't understand why this is here. Shouldn't this element be the `enum` element directly?

Comment (klokane) [inline test/fixtures/mson-enum.json]: I made it according to `tag` example in specs.

But You are right, it seem to be one more level there.

@zdne ?
https://github.com/refractproject/refract-spec/blob/master/namespaces/mson-namespace.md#examples

Comment (pksunkara) [inline test/fixtures/mson-enum.json]: That was supposed to be a top level example. Here `# User` would be the top level.

File: test/fixtures/mson-nontyped-object.json [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                  "element": "string",
+                  "meta": {
+                    "name": "zip"
+                  },
+                  "content": "34567"
+                }
+              ]
+            },
+            {
+              "element": "object",
```
Comment (pksunkara) [inline test/fixtures/mson-nontyped-object.json]: This should be an array. The parser will parse this as array.

**Comment by klokane** [comment]
Moved on shared branche

**Pull request closed without merging**